README.md
## MagicMirror Module: FlightsAbove

[![MagicMirror2](https://img.shields.io/badge/MagicMirror-2.2.2-lightgray.svg)](https://github.com/MichMich/MagicMirror)
[![DocStatus](https://inch-ci.org/github/E3V3A/MMM-FlightsAbove.svg?branch=master)](https://inch-ci.org/github/E3V3A/MMM-FlightsAbove)
[![GitHub last commit](https://img.shields.io/github/last-commit/E3V3A/MMM-FlightsAbove.svg)](https://github.com/E3V3A/MMM-FlightsAbove)
[![Maintenance](https://img.shields.io/badge/Maintained%3F-yes-green.svg)](https://GitHub.com/E3V3A/MMM-FlightsAbove/graphs/commit-activity)
[![Average time to resolve an issue](http://isitmaintained.com/badge/resolution/E3V3A/MMM-FlightsAbove.svg)](http://isitmaintained.com/project/E3V3A/MMM-FlightsAbove "Average time to resolve an issue")

Display what is currently flying above your location


| STATUS: | Version | Date | Maintained? |
|:------- |:------- |:---- |:----------- |
| Working | `1.0.4` | 2018-03-13 | YES |


#### What is this module doing?

*MMM-FlightsAbove* is a [MagicMirror](https://github.com/MichMich/MagicMirror) module for displaying what is 
currently flying around in the airspace above some location. The *location* is defined by the *latitude* and 
*longitude* of your location, and a radius in *kilometers*, which determine the maximum distance to which you 
want to detect aircraft.

However, because maps you see and use are flat and squared, internally the application is using a 
*boundary box* (BB) to determine the area of interest. The BB sides are just the latitutes and longitudes of the map.
So to easily calculate the BB values given a geographical location, we have included a tool that does the calculation for you. 
Use this to copy/paste your BB into your configuration file. 

The flight data is provided by [FlightRadar24](https://www.flightradar24.com/), but could be easily extended 
to use other sources as well.

The specific flight data items available from the API that you can display are: 


| Data Item     | Description |
|:------------- |:----------- |
| id            | Unique F24 id |
| modes         | Mode-S Transponder code: 6-digit [hex] is a 24-bit ICAO issued code |
| latitude      | in decimal degrees |
| longitude     | in decimal degrees |
| bearing       | in [degrees] |
| altitude      | in [feet]  |
| speed         | in [knots] |
| squawk        | Mode-3/A Transponder code: 4-digit [octal] is the "Squawk" ATC assigned code |
| radar         | F24 "radar" data source (ADS-B, MLAT, RADAR (ATC feed), FLARM, "ESTIMATED") designator |
| model         | ICAO Aircraft Type Designator |
| registration  | ICAO Aircraft (tail) registration number |
| timestamp     | Timestamp [Unix/POSIX/epoch] |
| origin        | Departure Airport IATA |
| destination   | Destination Airport IATA |
| flight        | IATA Flight Id |
| ground        | Airplane is "onGround": [0/1]  |
| climb         | "Rate of Climb" (RoC) is a vertical speed [ft/m] |
| callsign      | ICAO ATC call signature |
| glider        | Aircraft is a glider:  [0/1]  (There are usually no or very few gliders registered on earth!) |



### Example Screenshots


Module Screenshot:

![Full](./images/FlightsAbove1.png)

(From the [MMM-Tabulator](https://github.com/E3V3A/MMM-Tabulator) Demo.)

Here is a screenshot when using the config option: `homeIata: "VNO"` and `compassHeading: true` to 
highlight flights with the origin or destination airport given by the IATA code "VNO", and show the 
compass heading in text form and not in degrees. 

![Full](./images/FlightsAbove2.png)

Using the config option `footerLegend: true` will enable a footer with the color legend.

![Full](./images/FlightsAbove_footer.png)


The `MagicRadarBB.html` *Boundary Box Calculator* Screenshot:

![Full](./images/MagicRadarBB.png)


---


### Dependencies

This module depend on the following *npm* packages:

* [jquery](https://github.com/jquery/jquery)  - is a fast, small, and feature-rich JavaScript library
* [jquery-ui](https://github.com/jquery/jquery-ui)  - for user interface interactions, effects, widgets, and themes
* [tabulator](https://github.com/olifolkerd/tabulator) - for an overkill use of JSON to HTML table generation
* [flightradar24-client](https://github.com/derhuerst/flightradar24-client/) - the API used to get FR24 flight data

These are also listed in the `package.json` file and should be installed automatically when using *npm*.
However, those may require other packages. 

### Compatibility

This module is compatible with other **remote control** modules such as:

* [Hello-Lucy](https://github.com/mykle1/Hello-Lucy) (This is a plug-in to MMM-voice.) 
* [MMM-voice](https://github.com/fewieden/MMM-voice)

Those allow you to control this module by Voice commands.
For example, in Hello-Lucy you can say: "`SHOW FLIGHTS`" or "`HIDE FLIGHTS`" to show or hide this module.


---


### Installation

Manual Installation:

```bash
cd ~/MagicMirror/modules
git clone https://github.com/E3V3A/MMM-FlightsAbove.git
cd MMM-Tabulator
```

To install the depdendencies: *jquery*, *jquery-ui* and *tabulator*, do:

```bash
npm install jquery --save
npm install jquery-ui-dist --save
npm install jquery.tabulator --save

```

To install the *flightradar24-client* (API), use:

```bash
npm install flightradar24-client
```


Alternatively, on a *unix based distribution, you can try to install all the dependencies with the Bash script:

```bash
chmod 755 install_deps.sh
./install_deps.sh
```


#### Easy Installation [**ToDo**]:

```bash
npm install mmm-flightsabove
```

---


### Configuration 

To configure the FlightsAbove, you need to do the following:

1. Add the Module to the global MM `config.js` 
2. Edit the global config to add the 4 lat/lon values for the Boundary Box of your location
3. [optional] Edit `MMM-FlightsAbove.js` Tabulator settings to:
    - select & configure the table rows you want to show
    - edit the column formatting to be used
    - select the default Tabulator CSS
4. [optional] Edit `MMM-FlightsAbove.css` for your own CSS taste


Add the module to the modules array in the `config/config.js` file by adding the following section. 

```javascript
{
    module: 'MMM-FlightsAbove',
    header: 'Flights Above',
    position: 'top_left',
    config: {
        header: "Flights Above",          // The module header text, if any. Use: "" to remove.
        compassHeading: true,             // Use a compass direction (NSWE) for the of heading indicator ("45" vs "NE")
        updateInterval: 180,              // [s] Radar scan/ping/update period in secodns [default 3 min]
        maxItems: 10,                     // MAX Number of planes to display [default is 10]
        // The geographical (map) Boundary-Box (BB), from within planes will be shown are given by:
        // the maximim Lat/Lon edges of: [N-lat, W-lon, S-lat, E-lon] - all in decimal degrees.
        radarBBox: [-8.20917,114.62177,-9.28715,115.71243], // "DPS" (Bali Airport) [default]
        //radarLocation: "23.2,54.2",     // [Lat,Lon] - The location of radar center in decimal degrees
        //radarRadius: 60,                // [km] - The maximum distance of planes shown.
        //watchList: "",                  // Alert or Highlight planes/flights/types on this watch list
    }
},
...
```

You can change this configuration later when you see that it works.

---

#### The Boundary Box Coordinates

Unfortunately, the API is operating with a geographical boundary-box (BB), rather than a sphercial radius from 
a given location. So you need to convert your locations Lat/Lon (in decimal degrees) to the 4 coordinates 
representing the sides of the BB. These are given by: `MaxLat, MinLon, MinLat, MaxLon`, or in other words, 
the `north, west, south, east` edges. 


However, I have provided a simple GoogleMap based tool that make this calculation trivial. 
Look at the picture above, to see what to expect. To use the tool the first time, do the following:

1. Have your `Google Maps API Key` handy. You can get one [HERE](https://developers.google.com/maps/documentation/javascript/get-api-key).
2. Edit the file: `./tools/MagicRadarBB.html` and replacing the text "YOUR_GOOGLE_MAPS_API_KEY" with your key.
3. Simply point your browser to one of the following URLs:
   * from a local browser, open the file: `./tools/MagicRadarBB.html` in the module directory.
   * from a local or remote browser, go to: http://192.168.xx.yy:8080/modules/MMM-FlightsAbove/tools/MagicRadarBB.html
4. Enter the *Latitude* and *Longitude* in decimal degrees, for your location.
5. Enter the *radius* in kilometer, out to which you want to see flights. 
6. [*optional*] You can also drag the central marker to any point on the map. (Use `+`/`-` to zoom in/out.)
7. Hit `Calculate` and copy the resulting coordinate text from the `FR24 BB` textbox into your MM config file.
8. If you get a blank map, your API key is probaly expired, invalid or over-used.

Eventually, we would like to calculate this automatically in the `node_helper.js` script. (See issues)


**Q: What radius should I use?**

**A:** Since there is little use in displaying planes you cannot see, what you really should be 
asking is, how far can I see an airplane? The answer obviously depend on where you are, and especially 
how high you are. If you are down in a valley surrounded by mountains, you can use your trigonometry 
skills to figure that its probably not much more than 30 km, *line of sight* (LOS). But if you are on 
top of a mountain with free horizon, we could be talking about 120 km, at which point the aircraft would 
just present itself as a tiny spot in the sky. This module is preset at a moderate 60 km LOS.

For an interesting London discussion, see [this](https://www.metabunk.org/how-far-away-can-i-see-an-aircraft.t4038/).


#### Configuration Options 


| Option            | Default          | Description  |
|:----------------- |:---------------- |:------------ | 
| header            | "Flights Above"  | This Module's header text |
| compassHeading    | true             | Use a compass direction (NSWE) for the of heading indicator ("45" vs "NE") |
| updateInterval    | 180              | Module data update rate in [seconds] |
| maxItems*         | 10               | MAX number of planes (table rows) to show |
| radarBBox         | [see text]       | The geographical boundary-box in NW-SE notation [dec.degrees] |
| radarLocation*    | [see text]       | The location of radar center in decimal degrees |
| radarRadius*      | [see text]       | The maximum distance of planes shown |
| watchList*        | [see text]       | Highlight planes/flights/types on watch list |
| speedunit*        | "kmh"            | [kmh,kn] The speed unit |
| altunit*          | "m"              | [m,ft] The Altitude unit | 
| fontsize*         | "small"          | [small,medium,large] |
| itemlist*         | [see text]       | List of items (columns) to be shown |

`*` - **not yet implemented - ToDo!**

The default table columns shown, are: `[flight,callsig,to,from,alt,speed,bearing,alt,model,squawk]`
But there are also: `[id,registration,model,modes,radar]` available. 


```diff
-▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃
```

> :warning: Please do not use an **updateInterval** shorter than ~2 minutes (120 seconds). A too short 
> setting is useless as planes do not move across the visible sky that fast, and no, you will not be 
> able to see F16's or MIG's, as they are not tracked by F24. So please respect the free services 
> provided by the Flightradar24 API. You also risk getting blocked by their servers if you rapid-fire 
> requests. Finally, the services provided by them could break at any time, there are no service guarantees. :warning: 

```diff
-▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃
```


#### *Tabulator* Configuration

The main part to configure is found under the `loadTabulator()` function:

```javascript
...
    flightTable.tabulator({
        height:flightTableHeight,           // [px] Set MAX height of table, this enables the Virtual DOM and improves render speed
        layout:"fitDataFill",               // Resize columns to fit thier data and ensure rows takeup the full table width
        //layout:"fitColumns",              // Resize columns so that they fit perfectly inside the width of the container
        layoutColumnsOnNewData:true,        // Adjust the column width to the data each time you load it into the table
        //headerSort:false,                 // Disable header sorter
        resizableColumns:false,             // Disable manual column resize
        responsiveLayout:true,              // Enable responsive layouts
        placeholder:"Waiting for data...",  // Display message to user on empty table
        initialSort:[                       // Define the sort order:
            {column:"altitude",     dir:"asc"},     // 1'st
            //{column:"flight",     dir:"desc"},    // 2'nd
            //{column:"bearing",    dir:"asc"},     // 3'rd
        ],
        columns:[
            ...
            {title:"Time", field:"timestamp", headerSort:false, sortable:false, visible:false, responsive:1, formatter:"ep2time"},
...

```

For further details about configuring the many Tabulator options, see [MMM-Tabulator](https://github.com/E3V3A/MMM-Tabulator/) and references therein.


---

#### Bugs and Warnings

:information_source: The data provided by the *flightradar24-client*, is not JSON, even if it resembles it. It is 
actually an *Array of JavaScript objects*. For an example of the raw data, look at the `demo.json` file.

For other bugs, issues, details and updates, please refer to the [issue tracker](https://github.com/E3V3A/MMM-FlightsAbove/issues).


#### Contribution

Feel free to post issues and PR's related to this module.  
For all other or general questions, please refer to the [MagicMirror Forum](https://forum.magicmirror.builders/).

#### Credits 

Most grateful thanks to:
* [raywo](https://github.com/raywo/) - for clarifying and fixing the API response mechanism 
* [tbbear](https://github.com/tbbear/) - for helping and clarifying JS Promises
* [fewieden](https://github.com/fewieden) - for helping out in forums and fixing issues


#### License 

[![GitHub 
license](https://img.shields.io/github/license/E3V3A/MMM-FlightsAbove.svg)](https://github.com/E3V3A/MMM-FlightsAbove/blob/master/LICENSE) 

(MIT)


CHANGELOG.md
### MMM-FlightsAbove Change Log

All notable changes to this project will be documented in this file.

**Q**: *Why keep a changelog?*  
**A**: [BECAUSE](http://keepachangelog.com/en/1.0.0/)

**Q**: *How?*  
**A**: By carefully describing what was: *added, fixed*, and *changed*.

---

#### [1.0.4] - 2018-03-13

- Added remote control of module from other modules (fixes #18)
- Added config option for squawk watchList (fixes #13)
- Added config option for homeIata (fixes #19)
- Fixed CSS font color bug (fixes #17)
- Fixed CSS table header background color
- Added config.txt template to copy into config.js
- Added new screenshot for flight table 
- Added badges to README for:
  + MagicMirror version
  + Documentation staus
  + Maintenance status
  + Average Issue resolve time
  + MIT License 
- Removed version info from node_helper.js and MMM-FLightsAbove.js
- Changed node_helper to use the custom installed flightradar24-client in `./lib` 
  until new API items are resolved.


#### [1.0.3] - 2018-03-04

- Added config option maxItems (fixes #4)
- Fixed immediate data avilablity for new clients (fixes #14)


#### [1.0.2] - 2018-03-01

- Added config option radarBBox (fixes #7)
- Added config option compassHeading (fixes #12)
- Fixed Bearing center alignment
- Added layoutColumnsOnNewData to Tabulator
- Enabled fitDataFill in Tabulator
- Increased default MAX table height


#### [1.0.1] - 2018-03-01

- Fixed update timer for multiple connection instances (fixes #10)
- Fixed updateInterval config to use seconds instead of milliseconds (fixes #3)

#### [1.0.0] - 2018-02-28

- First working Version
- Fixed radarPing in node_helper (@raywo)
- Removed debug logging from module
- Update README with new info
- Added MagicRadarBB calculator


#### [0.0.1] - 2018-02-27

- Initial (non-working) Version


node_helper.js
/* --------------------------------------------------------------------------
 * Module:       MMM-FlightsAbove
 * FileName:     node_helper.js
 * Author:       E:V:A
 * License:      MIT
 * Date:         2018-03-13
 * Version:      1.0.4
 * Description:  A MagicMirror module to display planes in the sky above you
 * Format:       4-space TAB's (no TAB chars), mixed quotes
 *
 * URL:          https://github.com/E3V3A/MMM-FlightsAbove
 * Template:     https://github.com/E3V3A/MMM-Tabulator
 * --------------------------------------------------------------------------
 * Tabulator Requires:
 *      "node_modules/jquery/dist/jquery.min.js",
 *      "node_modules/jquery-ui-dist/jquery-ui.min.js",
 *      "node_modules/jquery-ui-dist/jquery-ui.css",
 *      "node_modules/jquery.tabulator/dist/js/tabulator.js",
 *      "node_modules/jquery.tabulator/dist/css/tabulator.css",
 *
 * flightradar24-client Requires:
 *      "dot-prop"
 *      "fetch-ponyfill"
 *      "moment-timezone"
 *      "parse-jsonp"
 *      "pinkie-promise"
 * --------------------------------------------------------------------------
 * DEV NOTES:
 *   - Try-catch is only for synchronous operations! (not async)
 *   -  Google maps BoundingBox Limits are defined by 2 corner points:
 *      SW and NE, whereas Flightradar24 API are using NW and SE corners!
 *      radar(LatMax/North, LonMin/West, LatMin/South, LonMax/East)
 *   - .
 * --------------------------------------------------------------------------
 */

/*jslint node: true */
'use strict';

const NodeHelper = require("node_helper");
const fs = require('fs');
//const radar = require('flightradar24-client/lib/radar'); // Returns an Array of JS objects, not JSON!
const radar = require('./lib/radar'); // Returns an Array of JS objects, not JSON!

module.exports = NodeHelper.create({

    start: function() {
        console.log(this.name + " started");
        this.config = null;
        this.started = false;   // (update timer)
        // get radar CONFIG here?
        //this.sendSocketNotification("REQUEST_RADAR_CONFIG");
    },

    stop: function() {
        console.log(this.name + " is shutting down");
    },

    radarPing: function() {
            const bb = this.config.radarBBox
            radar(bb[0],bb[1],bb[2],bb[3])
            .then((flights) => {
                this.radarData = flights; // save data for new clients
                this.sendSocketNotification("NEW_RADAR_DATA", this.radarData);
                //console.log("Sending NEW_RADAR_DATA:");
                //console.log(flights);
            })
            .catch(function(error) {
                console.log(error);
            });
    },

    socketNotificationReceived: function (notification, payload) {
        // We save this.started (update timer) status, to prevent mutiple timer restarts
        // for each new client connection/instance.
        const self = this; // (update time)
        if (notification === "START_RADAR" && this.started == false) {
            //console.log("Received START_RADAR");
            this.config = payload;
            this.radarPing();
            setInterval(() => {
                this.radarPing();
            }, this.config.updateInterval * 1000 );
            self.started = true; // update timer is started
        } else if (notification === "START_RADAR" && this.started == true) {
            this.sendSocketNotification("NEW_RADAR_DATA", this.radarData);
        }
    }

});


MMM-FlightsAbove.js
/* --------------------------------------------------------------------------
 * Module:       MMM-FlightsAbove
 * FileName:     MMM-FlightsAbove.js
 * Author:       E:V:A
 * License:      MIT
 * Date:         2018-03-13
 * Version:      1.0.4
 * Description:  A MagicMirror module to display planes in the sky above you
 * Format:       4-space TAB's (no TAB chars), mixed quotes
 *
 * URL1:         https://github.com/E3V3A/MMM-FlightsAbove
 * --------------------------------------------------------------------------
 * About Tabulator CSS & Themes:
 *   http://tabulator.info/docs/3.4?#css
 *   http://tabulator.info/examples/3.4?#theming
 * --------------------------------------------------------------------------
 * Tabulator Requires:
 * "node_modules/jquery/dist/jquery.min.js",
 * "node_modules/jquery-ui-dist/jquery-ui.min.js",
 * "node_modules/jquery-ui-dist/jquery-ui.css",
 * "node_modules/jquery.tabulator/dist/js/tabulator.js",
 * "node_modules/jquery.tabulator/dist/css/tabulator.css",
 * --------------------------------------------------------------------------
 */

'use strict'

Module.register('MMM-FlightsAbove',{

    defaults: {
        header: "Flights Above",            // The module header text, if any. (Use: "" to remove.)
//      footerLegend: true,                 // Enable the color-code legend to the table footer
        compassHeading: false,              // Type of heading indicator. ["true" gives "NE", instead of "45" (degrees)]
        maxItems: 8,                        // MAX Number of rows (flights) to display [default is 8]
        homeIata: "DPS",                    // Airport IATA code for highlighting flights To/From from a "home" airport
//        radarLocation: "-8.748,115.167",  // [Lat,Lon] - The location of radar center in decimal degrees
//        radarRadius: 60,                  // [km] - The maximum distance of planes shown.
        radarBBox: [-8.20917,114.62177,-9.28715,115.71243],  // [NLat,WLon,SLat,ELon] in (dec degrees) Default: "DPS"@60 km
        watchList: [1276,5000,5400,5600,6000,6100,6400],     // Highlight planes/flights/types on watch list
        updateInterval: 180                 // [seconds] Radar scan/ping/update period [default 3 min = 180 seconds]
    },

    requiresVersion: "2.1.0",

    start: function() {
        this.loaded = false;
        // This should be CONFIG!
        //this.sendSocketNotification("CONFIG", this.config);
        this.sendSocketNotification("START_RADAR", this.config);
    },

    getDom: function() {
        let w = document.createElement("div");  // Let w be the "wrapper"
        w.id = "flightsabove";                  // The id used by Tabulator

        if (!this.loaded) {
            w.innerHTML = "Loading...";
            w.className = "dimmed light small";
            return w;
        }
        if (!this.data) {
            w.innerHTML = "No data!";
            w.className = "dimmed light small";
            return w;
        }
        w.innerHTML = "Waiting for Tabulator...";
        return w;
    },

    getScripts: function() {
        return [
            this.file('node_modules/jquery/dist/jquery.min.js'),
            this.file('node_modules/jquery-ui-dist/jquery-ui.min.js'),
            this.file('node_modules/jquery.tabulator/dist/js/tabulator.js')
        ];
    },

    getStyles: function() {
        return [
            this.file('node_modules/jquery-ui-dist/jquery-ui.css'),
            //this.file('node_modules/jquery.tabulator/dist/css/tabulator.css'),                // Standard Theme
            this.file('node_modules/jquery.tabulator/dist/css/tabulator_midnight.min.css'),     // Midnight Theme
            //this.file('node_modules/jquery.tabulator/tabulator_simple.min.css'),              // Simple Theme
            //this.file('node_modules/jquery.tabulator/bootstrap/tabulator_bootstrap.min.css'), // Bootstrap Theme
            "MMM-FlightsAbove.css"                                                              // FlightsAbove Theme
        ];
    },

    getTranslations: function() {
        return false;   // Nothing to translate
    },

    // This come from the MM CORE or from other modules
    notificationReceived: function (notification, payload, sender) {
        if (notification === "DOM_OBJECTS_CREATED") {
            // The div with id "flightsabove" now exists, so we can load Tabulate.
            this.loadTabulate();
        }
        // This come from external control modules like: Hello-Lucy or MMM-voice.
        if (notification === 'HIDE_FLIGHTS') {
            this.hide(1000);
        }  else if (notification === 'SHOW_FLIGHTS') {
            this.show(1000);
        }
    },

    // This comes from YOUR module, usually "node_helper.js"
    socketNotificationReceived: function(notification, payload) {
        switch (notification) {
            //case "RADAR_CONFIG":
                //break;
            case "NEW_RADAR_DATA":
                this.loaded = true;
                this.setTableData(payload);
                //console.error("NEW_RADAR_DATA Received");
                //console.error(payload);
                break;
            default:
                console.error("Unmatched Notification: ", notification);
                //Log.error("Unmatched Notification: ", notification);
        }
    },

    //===================================================================================
    //  From here:  Tabulator specific code
    //===================================================================================
    loadTabulate: function() {

        // see: http://tabulator.info/docs/3.3#mutators
        /*Tabulator.extendExtension("mutator", "mutators", {
            ft2met:function(value, data, type, mutatorParams){
                return (value * 0.3048).toFixed(0);
            },
        });*/

        let self = this;
        Tabulator.extendExtension("format", "formatters", {
            ft2mt: function(cell, formatterParams) {              // Feet to Meters
                return  (0.3048*cell.getValue()).toFixed(0);
            },
            kn2km: function(cell, formatterParams) {              // Knots to Kilometers
                return  (1.852*cell.getValue()).toFixed(0);
            },
            ep2time: function(cell, formatterParams) {            // POSIX epoch to hh:mm:ss
                let date = new Date(cell.getValue());
                // We use "en-GB" only to get the correct formatting for a 24 hr clock, not your TZ.
                return date.toLocaleString('en-GB', { hour:'numeric', minute:'numeric', second:'numeric', hour12:false } );
            },
            deg2dir: function(cell, formatterParams) {            // Heading [deg] to approximate Compass Direction
                if (!self.config.compassHeading) { return cell; }
                let val = Math.floor((cell.getValue() / 22.5) + 0.5);
                let arr = ["N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE", "S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW"];
                return arr[(val % 16)];
            },

        });


        let flightTable = $("#flightsabove");
        var flightTableHeight = ( this.config.maxItems * 33 + 33 );   // @12px font-size we have [~33 px/row]
        var sqCodesList = this.config.watchList;
        var homePort = this.config.homeIata;

        flightTable.tabulator({
            height:flightTableHeight,           // [px] Set MAX height of table, this enables the Virtual DOM and improves render speed
            layout:"fitDataFill",               // Resize columns to fit thier data and ensure rows takeup the full table width
            //layout:"fitColumns",                // Resize columns so that they fit perfectly inside the width of the container
            layoutColumnsOnNewData:true,        // Adjust the column width to the data each time you load it into the table
            //headerSort:false,                   // Disable header sorter
            resizableColumns:false,             // Disable manual column resize
            responsiveLayout:true,              // Enable responsive layouts
            placeholder:"Waiting for data...",  // Display message to user on empty table
            initialSort:[                       // Define the sort order:
                {column:"altitude",     dir:"asc"},     // 1'st
                //{column:"flight",     dir:"desc"},    // 2'nd
                //{column:"bearing",    dir:"asc"},     // 3'rd
            ],
            columns:[
                {title:"Flight",        field:"flight",         headerSort:false, sortable:false, responsive:0, align:"left"}, // , width:250},
                {title:"CallSig",       field:"callsign",       headerSort:false, sortable:false, visible:true, responsive:3},
                {title:"To",            field:"destination",    headerSort:false, sortable:false, responsive:0},
                {title:"From",          field:"origin",         headerSort:false, sortable:false, responsive:0},
                {title:"Speed",         field:"speed",          headerSort:false, sortable:false, responsive:2, formatter:"kn2km"},   // [km/h]
                {title:"Bearing",       field:"bearing",        headerSort:false, sortable:false, responsive:1, formatter:"deg2dir", align:"center"}, // [deg/dir]
                {title:"Alt [m]",       field:"altitude",       headerSort:false, sortable:false, responsive:0, formatter:"ft2mt", align:"right", sorter:"number"},
                //{title:"Alt [m]",       field:"altitude",       sortable:true,  responsive:0, align:"right", sorter:"number", mutateType:"data", mutator:ft2met"},
                // Additional items:
                {title:"F24id",         field:"id",             headerSort:false, sortable:false, visible:false},
                {title:"RegID",         field:"registration",   headerSort:false, sortable:false, visible:false},
                {title:"Model",         field:"model",          headerSort:false, sortable:false, visible:true,  responsive:1},
                {title:"ModeS",         field:"modes",          headerSort:false, sortable:false, visible:false},
//                {title:"ModeS",         field:"modeSCode",      headerSort:false, sortable:false, visible:false},
                {title:"Radar",         field:"radar",          headerSort:false, sortable:false, visible:false},
                {title:"Lat",           field:"latitude",       headerSort:false, sortable:false, visible:false},
                {title:"Lon",           field:"longitude",      headerSort:false, sortable:false, visible:false},
                {title:"Time",          field:"timestamp",      headerSort:false, sortable:false, visible:false, responsive:1, formatter:"ep2time"},

                {title:"RoC [ft/m]",    field:"climb",          headerSort:false, sortable:false, visible:false},
                {title:"Squawk",        field:"squawk",         headerSort:false, sortable:false, visible:true, responsive:1}, // formatter:"sqCheck"},
                {title:"isGND",         field:"ground",         headerSort:false, sortable:false, visible:false},
                {title:"isGlider",      field:"glider",         headerSort:false, sortable:false, visible:false},
//                {title:"RoC [ft/m]",    field:"rateOfClimb",    headerSort:false, sortable:false, visible:false},
//                {title:"Squawk",        field:"squawkCode",     headerSort:false, sortable:false, visible:true, responsive:1}, // formatter:"sqCheck"},
//                {title:"isGND",         field:"isOnGround",     headerSort:false, sortable:false, visible:false},
//                {title:"isGlider",      field:"isGlider",       headerSort:false, sortable:false, visible:false},
            ],

            // INSERT rowFormatter functions here!
            rowFormatter: function(row) {

                //let sqCodesMIL = [44xx,5000,5100-5300,5400,6000,6100,6400,7501-7577]; // Squawk codes for Military/Government operations
                //let sqCodesNATO = [74xx]; // Squawk codes for NATO flight operations
                const sqCodesEMG = [7100,7500,7600,7700];  // Squawk codes for Emergencies
                const sqCodesUFO = ["0000",7777];          // Squawk codes that are highly unusual and should not be assigned

                var data = row.getData(); //get data object for row

                // Highlight Flights From: "DPS" (airport IATA code)
                if(data.origin == homePort && data.flight != "null"){
                    row.getElement().css({"color":"#33D9FF", "font-weight":"bold"});    // bright-blue bold
                }

                // Highlight Flights To: "DPS" (aiport IATA code)
                if(data.destination == homePort && data.flight != "null"){
                    row.getElement().css({"color":"#0ff", "font-weight":"normal"});       // aqua normal
                }

                // Highlight Flights with Emergency squawks:
                for (let i of sqCodesEMG) {
                    if (data.squawk == i) {
                        row.getElement().css({"color":"red", "font-weight":"bold"});    // red bold
                        //console.log("Emergency Flight!", i);
                    }
                }

                // Highlight Flights with UFO squawks:
                for (let i of sqCodesUFO) {
                    if (data.squawk == i) {
                        row.getElement().css({"color":"#ffa500", "font-weight":"bold"});   // orange bold
                        //console.log("Illegal Squawk!", i);
                    }
                }

                // Highlight Flights with squawks on watchList:
                for (let i of sqCodesList) {
                    if (data.squawk == i) {
                        row.getElement().css({"color":"#f0f", "font-weight":"bold"});   // fuchsia bold
                        //console.log("Illegal Squawk!", i);
                        this.sendNotification("SHOW_ALERT", {
                            title: "FlightRadar",
                            message: "A flight on your watch list has been detected!",
                            timer: "5000"
                        });
                    }
                }

            },
            // END (rowFormatter)

            // Add the footer color legend
            footerElement:$("<div class='tabulator-footer'>" +
                                "<span style='color:#33D9FF;'>▉</span> = To  " +
                                "<span style='color:#0ff;'>▉</span> = From  &nbsp;" +
                                "<span style='color:#f00;'>▉</span> = Emergency  " +
                                "<span style='color:#f0f;'>▉</span> = watchList  " +
                                "<span style='color:#0f0;'>▉</span> = Mil / Gov  " +
                                "<span style='color:#ffa500;'>▉</span> = UFO" +
                             "</div>"),
            //END
        });

        $(window).resize(function () {
            flightTable.tabulator("redraw");
        });
    },

    setTableData: function(data) {
        $("#flightsabove").tabulator("setData", data);
    }

    // To immediately sort (programatically):
    // $("#flightsabove").tabulator("setSort", "altitude", "asc");

    //trigger AJAX load on "Load Data via AJAX" button click
    /*$("#ajax-trigger").click(function(){
        $("#flightsabove").tabulator("setData", "/sample_data/ajax/data.php");
    });*/

    //===================================================================================

});
